### ISSUE TITLE
Properly decouple reading and parsing, to support finer-grained control flows such as non-blocking IO

### ISSUE THREAD

**Issue by infinity0** [dbuenzli/xmlm#16]
This library only supports blocking I/O at the moment - one can only use a `channel` as a source, or a `unit -> byte` which implicitly has blocking semantics. When calling `input` this will automatically read the source, so that reading logic gets executed deeper in the stack, than some of the parsing logic.

This is a bit unclean and backwards. It would be good to properly decouple the reading and the parsing, so that the former happens strictly before the latter. Then one can have finer grained control over the reading, which one doesn't (safely) have at the moment with xmlm.

For example, here is an initial strawman proposal to kick off a concrete discussion:

```ocaml
type source = [ ... | `Buffer ]

(* first argument is either [Some string] or [None] for EOF *)
val input_append : string option -> input -> unit

(* return value is either [Some signal] or [None] for "can't consume any more"
   [Xmlm.eoi] works the same. *)
val input_consume : input -> signal option
```

This would be useful for streaming XML-based protocols like XMPP, where you might not want to block reading a particular socket. For example:

```ocaml
let i = make_input (*params*) `Buffer in
(* stuff *)
match read ~timeout:5 fd with
| TimedOut -> (* do something else, and then loop again *)
| Read s ->
  input_append (Some s) i;
  let handle () =
    match input_consume i with
    | Some signal -> (* handle it *); handle ();
    | None ->
      if Xmlm.eoi i then (* handle it *) else ();
  in
  handle ()
(* or the equivalent using Lwt.pick and Lwt_unix.timeout *)
```

Technically, it is perhaps possible to do this with a `Function` source that throws an exception on timeout - but it's unclear from looking at the API whether this is safe or not. In fact, from looking at the source code I can see that some mutable fields get set before `i.uchar i.i` gets called so in all likelihood it's not safe and you can't just hope to re-enter `input` after the exception is raised.

**Comment by dbuenzli** [comment]
> This would be useful for streaming XML-based protocols like XMPP, where you might not want to block reading a particular socket.

Sure for now you need to spawn a thread. 

In any case if you are interested in non-blocking API design I suggest you have a look at https://github.com/dbuenzli/nbcodec (`jsonm` and `uutf` implement this pattern). 

Doing this for `xmlm` would require breaking the API and reprogram the whole thing in CPS style. It's unlikely I'll do this before we get effects (if we do) in the language, which eschew the need for this monkey circus altogether. The current API wouldn't even need to be changed, but should be for efficiency reasons, see again the above link for what it would look like. 

Meanwhile you may want to checkout @aantron's [XML parser](https://github.com/aantron/markup.ml) maybe it provides you what you'd like *now* (@aantron I'm interested to know if this is the case).

**Comment by aantron** [comment]
Markup.ml does indeed provide streaming non-blocking parsing. It is implemented internally in CPS, though the only non-blocking interface it currently exposes is for Lwt. That is a thin wrapper over the CPS implementation. I'm not sure what @infinity0 needs, but if it's not Lwt, perhaps a non-Lwt interface can also be exposed.

**Comment by dbuenzli** [comment]
Thanks @aantron !

**Comment by infinity0** [comment]
Thanks for the links and info! Yes nbcodec seems to fit what I am talking about, its `` `Manual`` is like my `` `Buffer`` above, and its `` `Await | `Yield | `End | `Error`` fits the `None` / `Xmlm.eoi` cases that I mentioned. This is very similar to the representation in the Haskell pipes library, perhaps you know that already. :)

Right now I'm only doing preliminary research for something I'll probably need in a few months' time, so I don't know if CPS would be a good match for what I need. But in general I think I prefer to write push-based control flows rather than pull-based control flows, because it's easier to do `select(1)`-like behaviour (with timeouts) on multiple streams, yet still have the main non-IO part of the code be pure.

I'm unsure how to achieve this sort of thing with the markup.ml [API](https://aantron.github.io/markup.ml/Markup.ASYNCHRONOUS.html), there is a `next` function, but as with xmlm it's unclear how to add timeouts to this in a safe way.

**Comment by aantron** [comment]
@infinity0 I'm not sure what you mean by timeouts here.

Markup.ml is mostly a pull API. The most general way to create an Lwt input is like this:

```ocaml
(* f : unit -> char option Lwt.t *)

let input = Markup_lwt.fn f in
let signals = Markup_lwt.parse_html input in

(* Now you can do Markup_lwt.next signals, etc. *)

(* Can also use Markup.parse_html if you don't care about asynchronous
   error reporting. *)
```

See [`Markup_lwt.fn`](http://aantron.github.io/markup.ml/Markup.ASYNCHRONOUS.html#VALfn). 

Doing a read (e.g. `Markup_lwt.next signals`) from your parsing signal stream will cause zero or more calls to `f ()`, each of which should evaluate to an Lwt promise that you complete at your leisure, or not at all. The calls are sequenced, i.e. `f` isn't called again until the latest promise completes.

I actually doubt that using `char option Lwt.t`s (as opposed to `string option Lwt.t`s) is a good idea here, but I haven't really looked into that.

I'm open to changing the API in various ways, as I haven't yet actually had any feedback on `Markup_lwt`.

**Comment by aantron** [comment]
Err, replace `parse_html` with `parse_xml` above for your case.

**Comment by infinity0** [comment]
The [timeout example](https://ocsigen.org/lwt/dev/api/Lwt_unix) in the lwt documentation explains what I mean quite well -

> For example, consider that you have two sockets sock1 and sock2. You want to read something from sock1 or exclusively from sock2 and fail with an exception if a timeout of 1 second expires, without reading anything from sock1 and sock2, even if they become readable in the future.

I worry that for both xmlm and markup.ml in their current forms, even if I spawn a thread and do `Lwt.pick [Lwt_unix.timeout 1.0; read ()]` then even if the timeout occurs first, the read will still succeed and consume the stream later because there is no "cancel" semantics. This is problematic, because perhaps when the timeout happens, I want to change the way in which I do the read or the way in which I react to it. (It may be possible to achieve the same effect, by communicating with the reading thread, but this would require much more complexity, and it would have to occur *after* the read unblocks.)

**Comment by aantron** [comment]
> then even if the timeout occurs first, the read will still succeed and consume the stream later because there is no "cancel" semantics.

This is independent of Markup.ml. As far as I know, you should do

```ocaml
let f () =
  let timeout = try%lwt Lwt_unix.timeout 1.0 with Lwt.Timeout -> Lwt.return None in
  let byte = read () in (* Succeeds with char option *)
  let p = Lwt.pick [timeout, byte] in
  p

let signals = Markup_lwt.(fn f |> parse_xml) in
```

(There are many variations on this)

Then, if `timeout` raises its exception, Lwt cancels the `byte` promise, and this presumably cancels the I/O inside `read` – depending on how you implement it. Note that `read ()` does not use Markup.ml, instead it (indirectly) is used *by* Markup.ml, so you have complete control over what happens here.

The above is perhaps not fully correct, as `None` signals EOF to Markup.ml. You have other options: you can simply not ever resolve the promise `p` that you pass to Markup.ml, and stop using the signal stream. You can also pass an arbitrary exception through Markup.ml, by failing `p` with some exception. The read of the signal stream that required the I/O that returned `p` will then fail with that exception.

Markup.ml is just some kind of transformer from sequences of reads of `signals` to sequences of reads of the input stream. It is supposed to be completely independent of how the input stream is actually read, whether reading the input stream ever completes, whether those reads can be canceled, and exceptions raised by reading the input stream. It "forwards" all those outcomes through to the signal stream, but inserts the HTML/XML parsing logic in between.

I am not sure if this answers your concerns, as I'm not sure exactly what they are – perhaps you could provide some pseudocode? With the existing `Markup_lwt`, the most flexible thing you can do is return a promise created by `Lwt.task ()`. Then, you can have a completely arbitrary strategy for resolving, failing, or forgetting it.

**Comment by aantron** [comment]
Ah, I guess you want to be able to forward a cancelation on a promise returned by Markup.ml to the underlying input stream...

**Comment by aantron** [comment]
So, I don't think that the parser can decide, in general for all users, how to handle cancelation on the signal stream. However, you can implement cancelation with `next` using something like this (it's just a dirty sketch, not meant to be that elegant or to compile):

```ocaml
(* Cancel reading of the input stream as soon as there is a cancelation on the
   signal stream. *)

let next =
  let canceled = ref false in  (* In case of cancelation before first read. *)
  let last_promise = ref None in

  let read_input () =
    if !canceled then
      Lwt.fail Lwt.Canceled
    else
      let p = some_cancelable_reading_function () in
      last_promise := Some p;
      p
  in

  let signals = Markup_lwt.(fn read_input |> parse_xml) in

  fun () ->
    let p = Markup_lwt.next signals in
    let p = Lwt.protected p in
    Lwt.on_cancel p (fun () ->
      canceled := true;
      match !last_promise with
      | None -> ()
      | Some p -> Lwt.cancel p)
```

The `Lwt.protected` call is needed because IIRC `Markup_lwt` uses `Lwt.wait` to create the output promises. I suppose that could be changed to `Lwt.task`.

Note that it is normal to wrap Markup.ml calls in things like this, though the sketch above is especially ugly. If you have some kind of general cancelation semantics in mind, let me know – otherwise, I think any library would have to force the user to communicate between the signal stream and the input reader somehow.

I actually think this makes a library easier to understand, though I'm worried about composability when using more functions than just `next`.

Perhaps the `Markup_lwt` interface should have special handling of cancelation built in somehow.

Also, we may want to move the Markup.ml-related part of this discussion to the Markup.ml repo to avoid spamming @dbuenzli... but up to you.

**Comment by infinity0** [comment]
Apologies that I don't have any concrete pseudocode at the moment, perhaps I'll have some in a few months.

> I guess you want to be able to forward a cancelation on a promise returned by Markup.ml to the underlying input stream.

This is interesting, but I think isn't needed by my future use-case. Rather, continuing with the example before:

```ocaml
let signals = Markup_lwt.(fn f |> parse_xml) in
```

I understand how this matches the Lwt example I quoted, however (my fault) neither of these fully captures the semantics I had in mind. Rather, it's more inspired by `select(2)` and other event loops from imperative languages. I'd like to do something along these lines:

```ocaml
let read_streams = [...] in
let parsers = [...] in (* one for each stream *)
let loop () =
  match select read_streams ~timeout:10 with
  | Timeout -> (* run other stuff *)
    (*  potentially do something to the parsers,
        so that future reads react differently, since they succeeded "later than expected" *)
  | Read (bytes, stream_index) ->
    bytes |> (List.nth parsers stream_index) |> (* do stuff with the signals *)
in
loop ()
```

In other words, there are a few extra things that the original example did not express very well:

1. The timeout does not cancel the read forever, one can resume the read later (of the same stream).
2. Handling the timeout can affect the behaviour of later reads (on the same or different streams).

**Comment by aantron** [comment]
I don't see why you couldn't do that with current Markup.ml. It looks completely orthogonal to the parser.

This looks like a push-based process, so you will need to create some extra machinery: a byte buffer for each parser, so that the parser can read from it when someone actually asks the parser for a signal, and a reference to a promise to resolve (if the parser asked to read bytes when none were available).

If you need, you can eagerly "pump" the parsers by calling `next signals` to cause the parsing logic to happen "early," as soon as data is available, even if the parsers have no consumer. Just put the resulting signals into some kind of stream or list. This might help to drain your byte buffers.

You can do whatever you want on timeout, the parsers don't really know how they eventually get the data that they *do* end up seeing. You might have spliced streams, caused arbitrary side effects, etc. If you want to do something like that, you just have to [change how you resolve those stored promises the parser is waiting on/change the behavior of your function that creates the promises in the first place] accordingly. You might need extra mutable references to fds, closures, or whatever you will be changing, but all that is part of the machinery you provide, and invisible to Markup.ml.

The parser just sees the `char option Lwt.t` promises being resolved. It has no idea what the underlying mechanism is.

The interface might be slower than you like due to the `char option Lwt.t`, but that can be fixed, if necessary. Also, you might be more interested in a direct CPS-based interface, at least for the parser input.

**Comment by infinity0** [comment]
Ah, OK I think I see what you mean now. Originally I was worried that if I "pump" the parser, this could block in some cases (as is the case for xmlm I think), but using a `char option Lwt.t` it seems possible to avoid that. I missed that when I was skimming over the API docs at first, I'll have to read it in more depth later. Thanks!
